# Initial SQL Queries
The purpose of this notebook is to run some initial queries and data exploration of the mlb_pitches sql database created in kaggle_dataset_sql_construction.ipynb.

Installing some needed packages:

In [1]:
from sqlalchemy import create_engine
import pandas as pd
from importlib import reload
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
%config InlineBackend.figure_formats = ['retina']
%matplotlib inline

plt.rcParams['figure.figsize'] = (9, 6)
sns.set(context='notebook', style='whitegrid', font_scale=1.2)

Setting up the sqlalchemy enginer and running some test queries:

In [2]:
#First, creating an engine and then importing the various .csv files.
engine = create_engine('postgresql://patrickbovard:localhost@localhost:5432/mlb_pitches')

### Pitches:

In [11]:
query = '''
SELECT * 
FROM pitches
LIMIT 5
;
'''
pitch_df = pd.read_sql(query, engine)

pitch_df.head()

px     pz  start_speed  end_speed spin_rate            spin_dir  \
0  0.416  2.963         92.9       84.1  2305.052             159.235   
1 -0.191  2.347         92.8       84.1  2689.935  151.40200000000004   
2 -0.518  3.284         94.1       85.2  2647.972             145.125   
3 -0.641  1.221         91.0       84.0   1289.59  169.75099999999995   
4 -1.821  2.083         75.4       69.6  1374.569             280.671   

   break_angle  break_length  break_y      ax  ...  event_num  b_score  \
0        -25.0           3.2     23.7   7.665  ...          3      0.0   
1        -40.7           3.4     23.7  12.043  ...          4      0.0   
2        -43.7           3.7     23.7  14.368  ...          5      0.0   
3         -1.3           5.0     23.8   2.104  ...          6      0.0   
4         18.4          12.0     23.8 -10.280  ...          7      0.0   

          ab_id  b_count s_count  outs  pitch_num  on_1b on_2b  on_3b  
0  2.015000e+09      0.0     0.0   0.0        1.0    0.0   0.0    0.0  
1  2.015000e+09      0.0     1.0   0.0        2.0    0.0   0.0    0.0  
2  2.015000e+09      0.0     2.0   0.0        3.0    0.0   0.0    0.0  
3  2.015000e+09      0.0     2.0   0.0        4.0    0.0   0.0    0.0  
4  2.015000e+09      1.0     2.0   0.0        5.0    0.0   0.0    0.0  

[5 rows x 40 columns]

Pulling the pitch columns:

In [12]:
pitch_df.columns

Index(['px', 'pz', 'start_speed', 'end_speed', 'spin_rate', 'spin_dir',
       'break_angle', 'break_length', 'break_y', 'ax', 'ay', 'az', 'sz_bot',
       'sz_top', 'type_confidence', 'vx0', 'vy0', 'vz0', 'x', 'x0', 'y', 'y0',
       'z0', 'pfx_x', 'pfx_z', 'nasty', 'zone', 'code', 'type', 'pitch_type',
       'event_num', 'b_score', 'ab_id', 'b_count', 's_count', 'outs',
       'pitch_num', 'on_1b', 'on_2b', 'on_3b'],
      dtype='object')

In [18]:
pitch_df.type.value_counts()

S    3
B    2
Name: type, dtype: int64

Type gives ball or strike.

### Games:

In [55]:
query = '''
SELECT * 
FROM games
;
'''
game_df = pd.read_sql(query, engine)

game_df.head()

Unnamed: 0  attendance  away_final_score away_team        date  \
0           0     35055.0               3.0       sln  2015-04-05   
1           1     45909.0               1.0       ana  2015-04-06   
2           2     36969.0               2.0       atl  2015-04-06   
3           3     31042.0               6.0       bal  2015-04-06   
4           4     45549.0               8.0       bos  2015-04-06   

   elapsed_time         g_id  home_final_score home_team start_time  \
0         184.0  201500001.0               0.0       chn    7:17 PM   
1         153.0  201500002.0               4.0       sea    1:12 PM   
2         156.0  201500003.0               1.0       mia    4:22 PM   
3         181.0  201500004.0               2.0       tba    3:12 PM   
4         181.0  201500005.0               0.0       phi    3:08 PM   

     umpire_1B       umpire_2B        umpire_3B      umpire_HP  \
0  Mark Wegner    Marty Foster  Mike Muchlinski   Mike Winters   
1    Ron Kulpa    Brian Knight    Vic Carapazza  Larry Vanover   
2     Laz Diaz  Chris Guccione      Cory Blaser    Jeff Nelson   
3    Ed Hickox     Paul Nauert   Mike Estabrook    Dana DeMuth   
4   Phil Cuzzi   Tony Randazzo      Will Little    Gerry Davis   

           venue_name                    weather                wind  delay  
0       Wrigley Field          44 degrees, clear   7 mph, In from CF    0.0  
1        Safeco Field         54 degrees, cloudy       1 mph, Varies    0.0  
2        Marlins Park  80 degrees, partly cloudy  16 mph, In from CF   16.0  
3     Tropicana Field           72 degrees, dome         0 mph, None    0.0  
4  Citizens Bank Park  71 degrees, partly cloudy   11 mph, Out to RF    0.0

Pulling the game columns:

In [56]:
game_df.columns

Index(['Unnamed: 0', 'attendance', 'away_final_score', 'away_team', 'date',
       'elapsed_time', 'g_id', 'home_final_score', 'home_team', 'start_time',
       'umpire_1B', 'umpire_2B', 'umpire_3B', 'umpire_HP', 'venue_name',
       'weather', 'wind', 'delay'],
      dtype='object')

### Atbats:

In [15]:
query = '''
SELECT * 
FROM atbats
LIMIT 5
;
'''
ab_df = pd.read_sql(query, engine)

ab_df.head()

Unnamed: 0  inning  top         ab_id         g_id  p_score  batter_id  \
0           0     1.0  1.0  2.019000e+09  201900001.0      0.0     594777   
1           1     1.0  1.0  2.019000e+09  201900001.0      0.0     545361   
2           2     1.0  1.0  2.019000e+09  201900001.0      0.0     571506   
3           3     1.0  0.0  2.019000e+09  201900001.0      0.0     543257   
4           4     1.0  0.0  2.019000e+09  201900001.0      0.0     656305   

   pitcher_id stand p_throws      event  o  
0      571666     L        R     Flyout  1  
1      571666     R        R     Flyout  2  
2      571666     L        R  Groundout  3  
3      502239     L        R     Single  0  
4      502239     R        R     Flyout  1

Pulling the at bat columns:

In [16]:
ab_df.columns

Index(['Unnamed: 0', 'inning', 'top', 'ab_id', 'g_id', 'p_score', 'batter_id',
       'pitcher_id', 'stand', 'p_throws', 'event', 'o'],
      dtype='object')

### Players:

In [17]:
query = '''
SELECT * 
FROM players
LIMIT 5
;
'''
player_df = pd.read_sql(query, engine)

player_df.head()

id first_name   last_name
0  452657        Jon      Lester
1  425794       Adam  Wainwright
2  457435       Phil        Coke
3  435400      Jason       Motte
4  519166       Neil     Ramirez

Player columns are very simple, just an id, first_name, last_name.

## Project Goal:

FOr this project, I would like to know some of the main information on each pitch: 
- Player Info: Pitcher and Batter Names
- Situation Info: Outs, Count, Runners on Base (i.e. count and base-out state), Score (is the pitcher winning or losing?, Inning - all of these can influence how a pitcher is going to approach a hitter  
- Physical Pitch Data: 
    - Pitch type, speed, spin rate
    - Pitch Location (X/Y coordinates and Zone)
- Time series data - what was the previous one (or multiple?) pitches thrown by the pitcher?  While it isn't extremely rare, pitchers usually wouldn't throw the same pitch and location again and again, since the batter could key in on it


Ultimately, I am trying to predict two things:
- the pitch type that will be thrown (i.e. classification of pitch type).  I will need to create a "pitch arsenal" for each pitcher, to avoid predicting a pitch that they have no chance of throwing.  One way to simplify would be to classify pitches as Fastball, Off-Speed, or Breaking.  This would reduce complexity but take away some detail.  Something to consider.
- the location of the pitch *after* the pitch is selected.  The location is heavily dependent on the pitch, so once the pitch is selected, it can be used as a feature to determin location.  I see this done either through regression (i.e. x and y coordinates) or classification (which "chunk" of the strike zone the pitch will be in).

### Needed columns:

With that in mind, I'll want to pick out some key columns from the dataframes to put together in one table or view.  These would be:

- Pitch: px, pz, zone, type, pitch_type, event_num, b_score, ab_id, b_count, s_count, outs, pitch_num, on_1b, on_2b, on_3b
- Game: hone_team, away_team, date
- At Bats: inning, top (1/0 binary), p_score (pitcher score), batter_id, pitcher_id, stand (batter hand), p_throws (pitcher hand), event (pitch result)
- Player: first_name, last_name

### Table Links:

Based on the table structure these are linked as follows: 
- Pitch: contains ab_id to link to atbats
- At_bats: has g_id to link to game, batter_id and pitcher_id to link to player  
Thus, would need to connect at bats to both players and games, then link that join to pitches.

## Joins:

Starting to run some joins on the tables, as a proof of concept:

For joining game on atbat, there should be one row per at bat:

In [34]:
query = '''
SELECT a.inning, a.batter_id, a.pitcher_id, a.top, a.ab_id, a.p_score, a.stand, a.p_throws, a.event, a.o,g.date
FROM atbats as a
RIGHT JOIN games as g
    ON a.g_id = a.g_id
LIMIT 5
;
'''
test_df = pd.read_sql(query, engine)

test_df.head()

inning  batter_id  pitcher_id  top         ab_id  p_score stand p_throws  \
0     1.0     594777      571666  1.0  2.019000e+09      0.0     L        R   
1     1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
2     1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
3     1.0     543257      502239  0.0  2.019000e+09      0.0     L        R   
4     1.0     656305      502239  0.0  2.019000e+09      0.0     R        R   

       event  o        date  
0     Flyout  1  2015-04-05  
1     Flyout  2  2015-04-05  
2  Groundout  3  2015-04-05  
3     Single  0  2015-04-05  
4     Flyout  1  2015-04-05

So the above gives me a table with at-bat information combined with the data.  Next step is getting the player names in that same output.

In [58]:
query = '''
--Queuing up the game/at bat info from teh above query:
WITH game_ab AS (
SELECT a.inning, a.batter_id, a.pitcher_id, a.top, a.ab_id, a.p_score, a.stand, a.p_throws, a.event, a.o,g.date, g.home_team, g.away_team
FROM atbats as a
RIGHT JOIN games as g
    ON a.g_id = g.g_id
LIMIT 20
)

SELECT g.*, 
p.first_name as Pitcher_First_Name, p.last_name as Pitcher_Last_Name, 
h.first_name as Hitter_First_Name, h.last_name as Hitter_Last_Name

FROM game_ab as g

--first, joining up the pitcher's name:
LEFT JOIN players as p
    ON g.pitcher_id = p.id

--now, joining the hitter's name:
LEFT JOIN players as h
    ON g.batter_id = h.id

--Ordering and limiting:
ORDER BY g.ab_id ASC
LIMIT 20
;
'''
test_df = pd.read_sql(query, engine)

test_df.head(12)

inning  batter_id  pitcher_id  top         ab_id  p_score stand p_throws  \
0      1.0     594777      571666  1.0  2.019000e+09      0.0     L        R   
1      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
2      1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
3      1.0     543257      502239  0.0  2.019000e+09      0.0     L        R   
4      1.0     656305      502239  0.0  2.019000e+09      0.0     R        R   
5      1.0     572039      502239  0.0  2.019000e+09      0.0     R        R   
6      1.0     501981      502239  0.0  2.019000e+09      0.0     R        R   
7      2.0     592743      571666  1.0  2.019000e+09      0.0     R        R   
8      2.0     405395      571666  1.0  2.019000e+09      0.0     R        R   
9      2.0     446359      571666  1.0  2.019000e+09      0.0     R        R   
10     2.0     600303      571666  1.0  2.019000e+09      0.0     L        R   
11     2.0     595777      502239  0.0  2.019000e+09      0.0     L        R   

           event  o        date home_team away_team pitcher_first_name  \
0         Flyout  1  2019-03-28       oak       ana               Mike   
1         Flyout  2  2019-03-28       oak       ana               Mike   
2      Groundout  3  2019-03-28       oak       ana               Mike   
3         Single  0  2019-03-28       oak       ana             Trevor   
4         Flyout  1  2019-03-28       oak       ana             Trevor   
5       Forceout  2  2019-03-28       oak       ana             Trevor   
6         Flyout  3  2019-03-28       oak       ana             Trevor   
7      Groundout  1  2019-03-28       oak       ana               Mike   
8   Hit By Pitch  1  2019-03-28       oak       ana               Mike   
9        Pop Out  2  2019-03-28       oak       ana               Mike   
10        Flyout  3  2019-03-28       oak       ana               Mike   
11        Triple  0  2019-03-28       oak       ana             Trevor   

   pitcher_last_name hitter_first_name hitter_last_name  
0              Fiers              Kole          Calhoun  
1              Fiers              Mike            Trout  
2              Fiers            Justin             Bour  
3             Cahill            Robbie         Grossman  
4             Cahill              Matt          Chapman  
5             Cahill           Stephen         Piscotty  
6             Cahill             Khris            Davis  
7              Fiers         Andrelton          Simmons  
8              Fiers            Albert           Pujols  
9              Fiers              Zack           Cozart  
10             Fiers             Tommy        La Stella  
11            Cahill         Jurickson           Profar

Excellent, based on the [game summary](https://www.mlb.com/gameday/angels-vs-athletics/2019/03/28/566086#game_state=final,lock_state=final,game_tab=box,game=566086) these events match up perfectly with what happened in the Angels-A's game on 3/28/2019, so this query is acting as intended.

Final step is adding in the pitch information to the above query:

In [75]:
query = '''
--Queuing up the game/at bat info from teh above query:
WITH game_player_ab AS (

WITH game_ab AS (
SELECT a.inning, a.batter_id, a.pitcher_id, a.top, a.ab_id, a.p_score, a.stand, a.p_throws, a.event, g.date, g.home_team, g.away_team
FROM atbats as a
RIGHT JOIN games as g
    ON a.g_id = g.g_id
LIMIT 20
)

SELECT g.*, 
p.first_name as Pitcher_First_Name, p.last_name as Pitcher_Last_Name, 
h.first_name as Hitter_First_Name, h.last_name as Hitter_Last_Name

FROM game_ab as g

--first, joining up the pitcher's name:
LEFT JOIN players as p
    ON g.pitcher_id = p.id

--now, joining the hitter's name:
LEFT JOIN players as h
    ON g.batter_id = h.id

--Ordering and limiting:
ORDER BY g.ab_id ASC
LIMIT 20
)

SELECT 
--First, taking all the data above, along with the batting team score:
gpa.*, pi.b_score,

--Now, selecting data from the pitch table:
pi.px, pi.pz, pi.pitch_type, pi.start_speed,pi.type, pi.b_count, pi.s_count, pi.outs, pi.pitch_num,

--Finally, adding in previous pitch info (for that at bat).  This will lead to some NaN values for the first pitch of an at-bat, which is expected:
(max(pi.pitch_type) OVER (PARTITION BY gpa.ab_id ORDER BY pi.pitch_num ASC ROWS 1 PRECEDING EXCLUDE CURRENT ROW)) as last_pitch_type,
(max(pi.px) OVER (PARTITION BY gpa.ab_id ORDER BY pi.pitch_num ASC ROWS 1 PRECEDING EXCLUDE CURRENT ROW)) as last_pitch_px,
(max(pi.pz) OVER (PARTITION BY gpa.ab_id ORDER BY pi.pitch_num ASC ROWS 1 PRECEDING EXCLUDE CURRENT ROW)) as last_pitch_pz


FROM pitches as pi
RIGHT JOIN game_player_ab as gpa
    ON pi.ab_id = gpa.ab_id
LIMIT 20
;
'''
test_df = pd.read_sql(query, engine)

test_df.head(12)

inning  batter_id  pitcher_id  top         ab_id  p_score stand p_throws  \
0      1.0     594777      571666  1.0  2.019000e+09      0.0     L        R   
1      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
2      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
3      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
4      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
5      1.0     545361      571666  1.0  2.019000e+09      0.0     R        R   
6      1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
7      1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
8      1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
9      1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
10     1.0     571506      571666  1.0  2.019000e+09      0.0     L        R   
11     1.0     543257      502239  0.0  2.019000e+09      0.0     L        R   

        event        date  ... pitch_type start_speed type b_count s_count  \
0      Flyout  2019-03-28  ...         FF        88.8    X     0.0     0.0   
1      Flyout  2019-03-28  ...         FF        89.9    C     0.0     0.0   
2      Flyout  2019-03-28  ...         SL        85.7    S     0.0     0.0   
3      Flyout  2019-03-28  ...         CH        85.4    B     0.0     1.0   
4      Flyout  2019-03-28  ...         CH        84.6    B     1.0     1.0   
5      Flyout  2019-03-28  ...         FF        90.9    X     2.0     1.0   
6   Groundout  2019-03-28  ...         FT        89.0    B     0.0     0.0   
7   Groundout  2019-03-28  ...         CH        84.4    B     1.0     0.0   
8   Groundout  2019-03-28  ...         FF        90.1    B     2.0     0.0   
9   Groundout  2019-03-28  ...         FF        89.2    C     3.0     0.0   
10  Groundout  2019-03-28  ...         SL        86.4    X     3.0     0.0   
11     Single  2019-03-28  ...         SI        89.8    C     0.0     0.0   

   outs  pitch_num  last_pitch_type  last_pitch_px last_pitch_pz  
0   0.0        1.0             None            NaN           NaN  
1   1.0        1.0             None            NaN           NaN  
2   1.0        2.0               FF           0.34          2.31  
3   1.0        3.0               SL          -0.05          2.03  
4   1.0        4.0               CH           0.49          0.92  
5   1.0        5.0               CH          -0.13          1.11  
6   2.0        1.0             None            NaN           NaN  
7   2.0        2.0               FT           0.62          0.25  
8   2.0        3.0               CH          -1.55          2.72  
9   2.0        4.0               FF          -1.03          2.96  
10  2.0        5.0               FF          -0.48          3.16  
11  0.0        1.0             None            NaN           NaN  

[12 rows x 29 columns]

Again, based on [the same game summary](https://www.mlb.com/gameday/angels-vs-athletics/2019/03/28/566086#game_state=final,lock_state=final,game_tab=play-by-play,game=566086), this follows the game: Kole Calhoun flew out on a one pitch at-bat (88.8 mph four-seam fastball), then Mike Trout flew out after a 5 pitch at bat with a 2-1 count.  While I can't verify the exact pitch coordinates (Px, Py), visually they look close based on the game summary pitches.  Additionally, the previous pitch info is matching up well, so the query seems to be pulling data as intended.  

So far so good.